asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py

```diff
@@ -0,0 +1,530 @@
+
+new_handles = handles + [boundary_patch, hn_zone_patch]
+new_labels = labels + ["Plymouth LA boundary", "Existing heat network zones"]
+
+plt.legend(
+    handles=new_handles,
+    labels=new_labels,
+    loc="upper left",
+    bbox_to_anchor=(1.05, 1),
+    title="Layers",
+)
+
+plt.show()
+
+# %%
+# What % of residential UPRNs are in existing HNZ?
+prop_in_hn_zone = plymouth_uprns_hnz_labelled_df.select(
+    pl.col("in_hn_zone").mean()
+).item()
+
+print(f"Proportion in HN zone: {prop_in_hn_zone:.3f}")
+
+# %% [markdown]
+# #### **Investigating which spatial signature types are in existing heat network zones**
+
+# %%
+# Spatial signatures in Plymouth LA boundaries only
+spatial_signatures_plymouth_gdf = spatial_signatures_gb_simplified_gdf.sjoin(
+    plymouth_la_boundaries_gdf, how="inner", predicate="intersects"
+).drop(columns="index_right")
+
+# %%
+# colour for spatial signature types
+unique_types = spatial_signatures_plymouth_gdf["type"].unique()
+colors = plt.cm.tab20.colors
+color_dict = {t: colors[i % len(colors)] for i, t in enumerate(unique_types)}
+
+
+fig, ax = plt.subplots(figsize=(15, 7.5))
+
+# plot spatial signatures polygons with mapped colors
+spatial_signatures_plymouth_gdf.plot(
+    ax=ax, color=spatial_signatures_plymouth_gdf["type"].map(color_dict)
+)
+
+# overlay Plymouth LA boundaries
+plymouth_la_boundaries_gdf.plot(ax=ax, facecolor="none", edgecolor="black", linewidth=1)
+
+# overlay Plymouth heat network zones
+plymouth_hn_zones_gdf.plot(ax=ax, facecolor="none", edgecolor="blue", linewidth=1)
+
+# manual legend for spatial signature types
+legend_handles = [mpatches.Patch(color=color_dict[t], label=t) for t in unique_types]
+
+# lines for boundaries and heat network zones
+legend_handles += [
+    Line2D([0], [0], color="black", lw=2, label="Plymouth LA boundary"),
+    Line2D([0], [0], color="blue", lw=2, label="Existing heat network zones"),
+]
+
+ax.legend(
+    handles=legend_handles,
+    title="Spatial Signature Type",
+    loc="upper left",
+    bbox_to_anchor=(1.05, 1),
+)
+
+plt.show()
+
+# %% [markdown]
+# 05/12/25: There is no clear consistency in the spatial signature types in existing heat network zones.
+
+# %% [markdown]
+# Inspecting existing HNZ one at a time:
+
+# %%
+plymouth_hn_zones_with_labels_gdf = gpd.sjoin(
+    plymouth_hn_zones_gdf,
+    spatial_signatures_plymouth_gdf[["type", "geometry"]],
+    how="left",
+    predicate="within",
+)
+
+plymouth_hn_zones_with_labels_agg_gdf = (
+    plymouth_hn_zones_with_labels_gdf.groupby("ZoneID")["type"]
+    .apply(lambda x: list(x.dropna().unique()))
+    .reset_index()
+)
+
+
+# %%
+# helper function to visually inspect each HNZ
+def inspect_spatial_signatures_in_hn_zone(zone_id: str):
+    zone_gdf = plymouth_hn_zones_gdf[plymouth_hn_zones_gdf["ZoneID"] == zone_id]
+
+    just_text_raw = str(zone_gdf.Just.values[0])
+    just_text = "\n".join(textwrap.wrap(just_text_raw, width=90))
```

Comment:
```suggestion
    # Add HN zone notes 
    just_text_raw = str(zone_gdf.Just.values[0])
    just_text = "\n".join(textwrap.wrap(just_text_raw, width=90))
```
Something like this, because it's not clear from looking at it what `Just` is.